sotodlib/data/toast_frame_utils.py

```diff
@@ -20,26 +20,118 @@
+FIELD_TO_CACHE= {
+    #"timestamps" : TOD.TIMESTAMP_NAME,
+    "flags_common" : TOD.COMMON_FLAG_NAME,
+    "site_velocity" : TOD.VELOCITY_NAME,
+    "site_position" : TOD.POSITION_NAME,
+    "flags" : TOD.FLAG_NAME,
+    "signal" : TOD.SIGNAL_NAME,
+}
+
+
+def recode_timestream(ts, params, rmstarget=2 ** 10, rmsmode="white"):
+    """ts is a G3Timestream.  Returns a new
+    G3Timestream for same samples as ts, but with data
+    scaled and translated with gain and offset,
+    rounded, and with FLAC compression enabled.
+
+    Args:
+        ts (G3Timestream) :  Input signal
+        compresspar (bool or dict) :  if True, compress with default
+            parameters.  If dict with 'rmstarget' member, override
+            default `rmstarget`.  If dict with `gain` and `offset`
+            members, use those instead.
+        params (None, bool or dict) :  If None, False or an empty dict,
+             no compression or casting to integers.  If True or
+             non-empty dictionary, enable compression.  Expected fields
+             in the dictionary ('rmstarget', 'gain', 'offset', 'rmsmode')
+             allow overriding defaults.
+        rmstarget (float) :  Scale the iput signal to have this RMS.
+            Should be much smaller then the 24-bit integer range:
+            [-2 ** 23 : 2 ** 23] = [-8,388,608 : 8,388,608].
+            The gain will be reduced if the scaled signal does
+            not fit within the range of allowed values.
+        rmsmode (string) : "white" or "full", determines how the
+            signal RMS is measured.
+    Returns:
+        new_ts (G3Timestream) :  Scaled and translated timestream
+            with the FLAC compression enabled
+        gain (float) :  The applied gain
+        offset (float) :  The removed offset
+
+    """
+    if not params:
+        return ts, 1, 0
+    gain = None
+    offset = None
+    if isinstance(params, dict):
+        if "rmsmode" in params:
+            rmsmode = params["rmsmode"]
+        if "rmstarget" in params:
+            rmstarget = params["rmstarget"]
+        if "gain" in params:
+            gain = params["gain"]
+        if "offset" in params:
+            offset = params["offset"]
+    v = np.array(ts)
+    vmin = np.amin(v)
+    vmax = np.amax(v)
+    if offset is None:
+        offset = 0.5 * (vmin + vmax)
+        amp = vmax - offset
+    else:
+        amp = np.max(np.abs(vmin - offset), np.abs(vmax - offset))
+    if gain is None:
+        if rmsmode == "white":
+            rms = np.std(np.diff(v)) / np.sqrt(2)
+        elif rmsmode == "full":
+            rms = np.std(v)
+        else:
+            raise RuntimeError("Unrecognized RMS mode = '{}'".format(rmsmode))
+        if rms == 0:
+            gain = 1
+        else:
+            gain = rmstarget / rms
+            # If the data have extreme outliers, we have to reduce the gain
+            # to fit the 24-bit signed integer range
+            while amp * gain >= 2 ** 23:
```

Comment:
This while loop should be un-indented by one level, to catch the case of zero RMS, no offset, large data value.  Yes, that case is unlikely.